test/fixtures/gcs-fixture/src/main/java/fixture/gcs/MockGcsBlobStore.java

```diff
@@ -0,0 +1,118 @@
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package fixture.gcs;
+
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.rest.RestStatus;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class MockGcsBlobStore {
+
+    private final ConcurrentMap<String, BlobVersion> blobs = new ConcurrentHashMap<>();
+
+    record BlobVersion(String path, long generation, BytesReference contents, String uploadId) {}
+
+    BlobVersion getBlob(String path, Long ifGenerationMatch) {
+        final BlobVersion blob = blobs.get(path);
+        if (blob == null) {
+            throw new BlobNotFoundException(path);
+        } else {
+            if (ifGenerationMatch != null) {
+                if (blob.generation != ifGenerationMatch) {
+                    throw new GcsRestException(
+                        RestStatus.PRECONDITION_FAILED,
+                        "Generation mismatch, expected " + ifGenerationMatch + " but got " + blob.generation
+                    );
+                }
+            }
+            return blob;
+        }
+    }
+
+    BlobVersion updateBlob(String path, Long ifGenerationMatch, BytesReference contents) {
+        return updateBlob(path, ifGenerationMatch, contents, null);
+    }
+
+    BlobVersion updateBlob(String path, Long ifGenerationMatch, BytesReference contents, String uploadId) {
+        return blobs.compute(path, (name, existing) -> {
+            if (existing != null) {
+                if (ifGenerationMatch != null) {
+                    if (ifGenerationMatch == 0) {
+                        throw new GcsRestException(
+                            RestStatus.PRECONDITION_FAILED,
+                            "Blob already exists at generation " + existing.generation
+                        );
+                    } else if (ifGenerationMatch != existing.generation) {
+                        throw new GcsRestException(
+                            RestStatus.PRECONDITION_FAILED,
+                            "Generation mismatch, expected " + ifGenerationMatch + ", got" + existing.generation
+                        );
+                    }
+                }
+                return new BlobVersion(path, existing.generation + 1, contents, uploadId);
+            } else {
+                if (ifGenerationMatch != null && ifGenerationMatch != 0) {
+                    throw new GcsRestException(
+                        RestStatus.PRECONDITION_FAILED,
+                        "Blob does not exist, expected generation " + ifGenerationMatch
+                    );
+                }
+                return new BlobVersion(path, 1, contents, uploadId);
+            }
+        });
+    }
+
+    BlobVersion updateResumableBlob(String path, String uploadId, BytesReference newContents) {
```

Comment:
This doesn't look right to me. Resumable blobs are not visible to `GET` requests, and you can in principle have two different resumable uploads for the same path at the same time which should not conflict like this.

Admittedly it looks wrong in the existing code too, but we're making it more wrong here I think: a resumed upload needs to check the generation match at the point that the upload completes and becomes visible, not when deciding whether or not to start accepting new data.

Reply:
Ah yep, I misinterpreted the purpose of the upload ID. I've implemented resumable uploads properly now, it was a bit of a can of worms, but I tested it with a streaming upload and it appeared to work as expected.

I restored the if-generation-match to work as you suggested, though I'm not sure if we actually use it in the fixture. In any case it was trivial to support.